### PR TITLE
Merge remote-tracking branch 'upstream/dev' into feat-fast-sync

### PR DESCRIPTION
**Pull request by rafal-ch**
This PR integrates the recent changes from the `dev` branch into the `feat-fast-sync` branch.

### CODE REVIEW

**Review by afck — COMMENTED** [review]
~Looks good apart from `StateStoreRequest` and `SharedObject`: I vaguely remember that we didn't want to reintroduce them, but I don't know the details.~
(Edit: We _do_ need `StateStoreRequest` because the block proposer now stores its state again!)
(Edit 2: And `SharedObject` is needed for in-memory message deduplication.)

**Review by afck — APPROVED** [review]
Looks good! Could use another review, though.

**Review by Fraser999 — APPROVED** [review]

File: node/src/effect.rs [hunk 1]
```diff
@@ -76,7 +77,9 @@ use once_cell::sync::Lazy;
 use serde::{Deserialize, Serialize};
 use smallvec::{smallvec, SmallVec};
 use tokio::{sync::Semaphore, time};
-use tracing::error;
+#[cfg(not(feature = "fast-sync"))]
```
Comment (Fraser999) [inline node/src/effect.rs]: This should be removed.

Comment (rafal-ch) [reply node/src/effect.rs]: Fixed in https://github.com/casper-network/casper-node/pull/2417/commits/2b7c5ef9fdba9ab09d0587bfba42c5b9be0353a0

**Review by Fraser999 — APPROVED** [review]

**Comment by rafal-ch** [comment]
Merging manually because the `nctl-upgrade-test` will fail when trying to update to the node based on the `feat-fast-sync` branch.

**Review by marc-casperlabs — COMMENTED** [review]
Checked only storage related things, those LGTM.

**Pull request merged**